### PR TITLE
Recover suggestions to introduce named lifetime under NLL

### LINKED ISSUE #96157: [NLL] Loss of suggestion to introduce a named lifetime parameter
In the `src/test/ui/self/arbitrary_self_types_pin_lifetime_mismatch.base.rs` test, we give the suggestion to `consider introducing a named lifetime parameter and update trait if needed` under migrate mode. Under NLL,  the error message is imo clearer but we lose that suggestion.

### PR DESCRIPTION
**Pull request by marmeladema**
Fixes #96157

r? @jackh726 

Built on top of #96385 so only the second commit is relevant

### CODE REVIEW

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs [hunk 1]
```diff
@@ -145,84 +146,85 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
...
-                    let is_impl = matches!(&node, hir::Node::ImplItem(_));
-                    let generics = match node {
-                        hir::Node::Item(&hir::Item {
-                            kind: hir::ItemKind::Fn(_, ref generics, ..),
-                            ..
-                        })
-                        | hir::Node::TraitItem(&hir::TraitItem { ref generics, .. })
-                        | hir::Node::ImplItem(&hir::ImplItem { ref generics, .. }) => generics,
-                        _ => return,
+pub fn suggest_adding_lifetime_params<'tcx>(
```
Comment (marmeladema) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs:158]: Method moved as a free standing function

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs [hunk 2]
```diff
@@ -145,84 +146,85 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
             }
         }
 
-        self.suggest_adding_lifetime_params(sub, ty_sup, ty_sub, &mut err);
+        if suggest_adding_lifetime_params(self.tcx(), sub, ty_sup, ty_sub, &mut err) {
+            err.note("each elided lifetime in input position becomes a distinct lifetime");
+        }
```
Comment (marmeladema) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs:151]: Note was originally in `suggest_adding_lifetime_params` but by moving it here we can avoid having the note under NLL.
However if we do want the note, I can just move it back to where it comes from

File: compiler/rustc_borrowck/src/diagnostics/region_errors.rs [hunk 3]
```diff
@@ -694,4 +698,24 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
         }
     }
+
+    fn suggest_adding_lifetime_params(
+        &self,
+        diag: &mut Diagnostic,
+        sub: RegionVid,
+        sup: RegionVid,
+    ) {
+        if let (Some(sub), Some(sup)) = (self.to_error_region(sub), self.to_error_region(sup)) {
```
Comment (jackh726) [inline compiler/rustc_borrowck/src/diagnostics/region_errors.rs:708]: Nit: this function can use let else to reduce nesting and clean it up

File: compiler/rustc_borrowck/src/diagnostics/region_errors.rs [hunk 4]
```diff
@@ -694,4 +698,24 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+            let f = |reg| {
```
Comment (jackh726) [inline compiler/rustc_borrowck/src/diagnostics/region_errors.rs:709]: If the function is converted to let else, I don't think having this as a closure helps much.

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs (hunk 2 above)
Comment (jackh726) [reply compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs:151]: Hmm. The note is kind of redundant with the NLL error output, I guess.

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs [hunk 5]
```diff
@@ -145,84 +146,85 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
...
-                        | hir::Node::ImplItem(&hir::ImplItem { ref generics, .. }) => generics,
-                        _ => return,
+pub fn suggest_adding_lifetime_params<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    sub: Region<'tcx>,
+    ty_sup: &Ty<'_>,
+    ty_sub: &Ty<'_>,
+    err: &mut Diagnostic,
+) -> bool {
+    if let (
```
Comment (jackh726) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs:165]: Preexisting nit: let else?

**Review by jackh726 — CHANGES_REQUESTED** [review]
Some nits but overall LGTM

File: compiler/rustc_borrowck/src/diagnostics/region_errors.rs (hunk 3 above)
Comment (marmeladema) [reply compiler/rustc_borrowck/src/diagnostics/region_errors.rs]: Fixed

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs (hunk 5 above)
Comment (marmeladema) [reply compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs]: Yes, but changed to let/else.

File: compiler/rustc_borrowck/src/diagnostics/region_errors.rs (hunk 4 above)
Comment (marmeladema) [reply compiler/rustc_borrowck/src/diagnostics/region_errors.rs]: Done

**Pull request merged**